egs/wsj/s5/steps/libs/nnet3/xconfig/basic_layers.py

```diff
@@ -0,0 +1,787 @@
+        Descriptors, into a 'normalized form' derived from parsing them as
+        Descriptors, replacing things like [-1] with the actual layer names,
+        and regenerating them as strings.  We stored this when the object was
+        initialized, in self.descriptors; this function just copies them back
+        to the config.
+        """
+
+        for key, desc_str_dict in self.descriptors.items():
+            self.config[key] = desc_str_dict['normalized-string']
+
+    def convert_to_descriptor(self, descriptor_string, all_layers):
+        """Convenience function intended to be called from child classes,
+        converts a string representing a descriptor ('descriptor_string')
+        into an object of type Descriptor, and returns it. It needs 'self' and
+        'all_layers' (where 'all_layers' is a list of objects of type
+        XconfigLayerBase) so that it can work out a list of the names of other
+        layers, and get dimensions from them.
+        """
+
+        prev_names = xutils.get_prev_names(all_layers, self)
+        tokens = xutils.tokenize_descriptor(descriptor_string, prev_names)
+        pos = 0
+        (descriptor, pos) = xutils.parse_new_descriptor(tokens, pos, prev_names)
+        # note: 'pos' should point to the 'end of string' marker
+        # that terminates 'tokens'.
+        if pos != len(tokens) - 1:
+            raise xparser_error("Parsing Descriptor, saw junk at end: " +
+                            ' '.join(tokens[pos:-1]), self.str())
+        return descriptor
+
+    def get_dim_for_descriptor(self, descriptor, all_layers):
+        """Returns the dimension of a Descriptor object. This is a convenience
+        function used in set_configs.
+        """
+
+        layer_to_dim_func = \
+                lambda name: xutils.get_dim_from_layer_name(all_layers, self,
+                                                            name)
+        return descriptor.dim(layer_to_dim_func)
+
+    def get_string_for_descriptor(self, descriptor, all_layers):
+        """Returns the 'final' string form of a Descriptor object,
+        as could be used in config files. This is a convenience function
+        provided for use in child classes;
+        """
+
+        layer_to_string_func = \
+                lambda name: xutils.get_string_from_layer_name(all_layers,
+                                                               self, name)
+        return descriptor.config_string(layer_to_string_func)
+
+    def get_name(self):
+        """Returns the name of this layer, e.g. 'affine1'.  It does not
+        necessarily correspond to a component name.
+        """
+
+        return self.name
+
+    ######  Functions that might be overridden by the child class: #####
+
+    # child classes should override this.
+    def set_default_configs(self):
+        raise Exception("Child classes must override set_default_configs().")
+
+    # this is expected to be called after set_configs and before check_configs()
```

Comment:
is expected to be called -> is called in the constructor.
I think this default function definition is potentially dangerous, but I guess I could be OK with it.